usb: handle endpoint-sized EP0 IN transactions

masz-nordic opened the pull request:
Fixes #15507

masz-nordic commented:
@dmazzella @anddam Can you test whether this fixes the issue for you?

finikorg commented on subsys/usb/usb_device.c:
```diff
@@ -233,12 +233,22 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
+	usb_dc_ep_write(USB_CONTROL_IN_EP0, usb_dev.data_buf, chunk, &chunk);
+#endif
 	/*Always EP0 for control*/
-	usb_dc_ep_write(0x80, usb_dev.data_buf, chunk, &chunk);
```
Cannot you just use `usb_dc_ep_write(0x80, NULL, 0, NULL)` ?

masz-nordic replied:
No, this would result in ZLP only instead of 64B of descriptor + ZLP.

finikorg commented on subsys/usb/usb_device.c:
```diff
@@ -233,12 +233,22 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
 {
 	u32_t chunk = usb_dev.data_buf_residue;
 
+#if CONFIG_USB_NRF52840
```
You need to be more specific, reel_board does not need this hack

masz-nordic replied:
That's odd, there should be nothing board specific in this case. I will try to get hold of a reel_board and test it.

finikorg replied:
I also really do not like placing this code to common `usb_device.c`, any change this can be moved to device controller?

finikorg replied:
![reel_board](https://user-images.githubusercontent.com/3072104/56715071-47665480-673f-11e9-8fb0-2435122b0d53.png)

finikorg replied:
For reel_board looks OK

masz-nordic replied:
The issue exists on Windows only (7 & 10). I've tried reel_board and was able to reproduce it:

![reel_board_zlp](https://user-images.githubusercontent.com/42532799/56718129-7a5c1880-673e-11e9-98d9-da4db9c3a3bc.png)

finikorg commented on include/drivers/usb/usb_dc.h:
```diff
@@ -308,6 +308,25 @@ int usb_dc_ep_flush(const u8_t ep);
+ *
+ * @return 0 on success, negative errno code on fail.
+ */
+int usb_dc_ep_write_zlp(const u8_t ep, const u8_t *const data,
```
we have already construction which is used for usb_transfer()
```			if (!(trans->flags & USB_TRANS_NO_ZLP)) {
				usb_write(ep, NULL, 0, NULL);
			}
```

finikorg commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -1619,13 +1619,15 @@ int usb_dc_ep_flush(const u8_t ep)
 	struct nrf_usbd_ctx *ctx = get_usbd_ctx();
 	struct nrf_usbd_ep_ctx *ep_ctx;
 	u32_t bytes_to_copy;
+	u8_t flags = zlp ? NRFX_USBD_TRANSFER_ZLP_FLAG : 0;
```
Cannot you use here something like?

```
flags = (ep == 0x80) && (data_len == 64) NRFX_USBD_TRANSFER_ZLP_FLAG : 0;
```

masz-nordic replied:
No, because Windows expects a transfer without ZLP when `wLength == 64` and transfer with ZLP when `wLength > 64`.

finikorg replied:
So in this case length is 64, so we do not send ZLP?

finikorg replied:
OK, Windows asks length=255 despite we return 64 in the previous call, and it needs to know that this is the last packet, this can be the general case then.

masz-nordic replied:
Basically this is what Windows expects:

| Descriptor length | wLength given by host |        ZLP?        |
|:-----------------:|:---------------------:|:------------------:|
|         64        |           64          |         :x:        |
|         64        |          255          | :heavy_check_mark: |

finikorg commented on subsys/usb/usb_device.c:
```diff
@@ -233,12 +233,22 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
 	u32_t chunk = usb_dev.data_buf_residue;
 
+#if CONFIG_USB_NRF52840
+	if ((chunk == USB_MAX_CTRL_MPS) && (len > USB_MAX_CTRL_MPS)) {
```
If the chunk=64 but there are more data to be send it would send ZLP after first packet, should we check `usb_dev.data_buf_residue`

masz-nordic committed:
usb: add zero-length packet after MPS-sized IN transaction

Send ZLP when host asks for a bigger length and the last chunk
is wMaxPacketSize long, to indicate the last packet.

Signed-off-by: Marcin Szymczyk <Marcin.Szymczyk@nordicsemi.no>

masz-nordic commented:
@finikorg Thanks for pointing out an easier solution.

finikorg commented on subsys/usb/usb_device.c:
```diff
@@ -233,14 +233,28 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
+
+	/*
+	 * Windows expects a ZLP after the last IN transaction that was
+	 * EP-sized if wLength was bigger than EP size.
```
MPS - sized?

finikorg reviewed (approved):
LGTM with minor comment about EP-sized

jfischer-phytec-iot commented on subsys/usb/usb_device.c:
```diff
@@ -233,14 +233,28 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
+	 * EP-sized if wLength was bigger than EP size.
+	 */
+	if (!usb_dev.data_buf_residue && chunk == USB_MAX_CTRL_MPS
+	    && len > chunk) {
```
The comment above and the commit message are confusing and not correct. It seems there is a bug in our device stack. Please change the commit message and comment to something like the one from @finikorg "Send ZLP when host asks for the bigger length and the last chunk is wMaxPacketSize to indicate about the last packet.

See also:
_The Data stage of a control transfer from an endpoint to the host is complete when the endpoint does one of
the following:
• Has transferred exactly the amount of data specified during the Setup stage
• Transfers a packet with a payload size less than wMaxPacketSize or transfers a zero-length packet
(5.5.3 USB Specification Revision 2.0)_

_A control pipe may have a variable-length data phase in which the host requests more data than is contained
in the specified data structure ....... If the data structure is an exact multiple of wMaxPacketSize for the pipe, the function will return a zero-length packet to indicate the end of the Data stage. (8.5.3.2 USB Specification Revision 2.0)_

finikorg replied:
Agree here with @jfischer-phytec-iot

masz-nordic replied:
Fixed

jfischer-phytec-iot commented:
I guess it should be backported to 1.14, @carlescufi @finikorg ?
